src/gen-ts.ts

```diff
@@ -8,295 +8,184 @@
  * Google as part of the polymer project is also subject to an additional IP
  * rights grant found at http://polymer.github.io/PATENTS.txt
  */
-// Requires node >= 7.6
 
-import {generate} from 'escodegen';
-import {Analysis, Analyzer, Element, ElementMixin, Feature, FSUrlLoader, Method, PackageUrlResolver, Property} from 'polymer-analyzer';
-import {Function as ResolvedFunction} from 'polymer-analyzer/lib/javascript/function';
+import * as analyzer from 'polymer-analyzer';
 
-const isInTestsRegex = /(\b|\/|\\)(test)(\/|\\)/;
-const isTest = (f: Feature) =>
-    f.sourceRange && isInTestsRegex.test(f.sourceRange.file);
+import {closureTypeToTypeScript} from './closure-types';
+import * as ts from './ts-ast';
+import {serializeTsDeclarations} from './ts-serialize';
 
-// const declarationKinds = ['element', 'element-mixin', 'namespace',
-// 'function']; const isDeclaration = (f: Feature) =>
-// declarationKinds.some((kind) => f.kinds.has(kind));
-
-const header = `declare namespace Polymer {
-  type Constructor<T> = new(...args: any[]) => T;
-`;
-const footer = `}`;
-
-export function generateDeclarations(root: string): Promise<string> {
-  const analyzer = new Analyzer({
-    urlLoader: new FSUrlLoader(root),
-    urlResolver: new PackageUrlResolver(),
+/**
+ * Analyze all files in the given directory using Polymer Analyzer, and return
+ * a TypeScript declarations document string that describes its contents.
+ */
+export async function generateDeclarations(rootDir: string): Promise<string> {
+  const a = new analyzer.Analyzer({
+    urlLoader: new analyzer.FSUrlLoader(rootDir),
+    urlResolver: new analyzer.PackageUrlResolver(),
   });
-
-  return analyzer.analyzePackage().then(generatePackage);
+  const analysis = await a.analyzePackage();
+  const ast = analyzerToAst(analysis);
+  return serializeTsDeclarations(ast);
 }
 
-function generatePackage(pkg: Analysis): Promise<string> {
-  const declarations = [header];
-
-  for (const feature of pkg.getFeatures()) {
-    // if (feature.sourceRange == null) {
-    //   continue;
-    // }
-    if (isTest(feature)) {
+/**
+ * Make a TypeScript declarations document from the given Polymer Analyzer
+ * result.
+ */
+function analyzerToAst(analysis: analyzer.Analysis): ts.Document {
+  const root: ts.Document = {
+    kind: 'document',
+    members: [],
+  };
+
+  for (const doc of analysis.getFeatures({kind: 'js-document'})) {
+    // TODO This is a very crude exclusion rule.
+    if (doc.url.match(/test|demo/)) {
```

Comment:
Some elements have `test` or `demo` in their names, look for a trailing slash after `test` or `demo`

Reply:
Done.